### PR TITLE
Added Standalone Github Workflow & Updated to `neo` 3.6.2

### LINKED ISSUE #259: Better output for RocksDbSharp.Native TypeInitializationException
Using Express on Mac or Linux carries [additional installation requirements](https://github.com/neo-project/neo-express#additional-neo-express-requirements). We should provide additional information - and perhaps a trouble shooter - when express can't locate the native rocksdb library

### LINKED ISSUE #355: Update the github workflow to produce and distribute Neo Express as a standalone application
**Is your feature request related to a problem? Please describe.**
We need to ship the neo-express executable. The current workflow doesn't generate or publish the executables. 

Related to #268 

**Describe the solution you'd like**
Update the workflow file to generate and publish the executables here on GitHub

**Additional context**
@cschuchardt88 seems experienced enough to deal with this, so I suggest he works on it

### PR DESCRIPTION
**Pull request by cschuchardt88**
## Release Build

**Download Page** - https://github.com/cschuchardt88/neo-express/releases/tag/3.6.95-preview-gf56687ff2e

---

## Change Log

**Code**

- Update `*.csproj` references to version 3.6.2 for `neo` `Neo.Cryptography.MPT`  `Neo.Network.RPC.RpcClient` `RocksDBStore` `Neo.Plugins.RpcServer` `Neo.Consensus.DBFT`
- Changed `BinarySerializer.Serialize` parameters to reflect new version of `neo`
- Fixed all the warnings
- Added more detailed error for `rocksdb`

**Github Actions**

- Added standalone single file build on release _(Manual Run)_
- Added native libraries _(`rocksdb`)_
- Added `standalone.yml`
- Updated `releases.yml`
- Updated `test.yml`
- Updated action versions

#### Someone that has `osx-x64` & `arm64` please test by running commands:
```
./neoxp create
./neoxp run -s 1
```
---

### Windows

- **win-x64**
image

- **win-x86**
image

### Linux

- **WSL linux-x64**
image
- **OpenSuSE linux-x64**
image

Closes #355 
Closes #259 
Helps #268

### CODE REVIEW

**Comment by cschuchardt88** [comment]
Also need to test on machine with no dotnet installed on it.

**Comment by gsmachado** [comment]
@cschuchardt88 congrats on this PR! I've been waiting for this for such a LONG TIME. 👏  

I will be testing on `osx-arm64` and `osx-x64` sometime this weekend. 👍

**Comment by cschuchardt88** [comment]
@Liaojinghui any word on this?

**Comment by Liaojinghui** [comment]
> @Liaojinghui any word on this?

LGTM. But maybe others also want to review, I'll wait.

**Comment by cschuchardt88** [comment]
@lock9 can you review please

**Comment by lock9** [comment]
Hi,

I can confirm that it's working on osx-arm64.
One thing that I would change are the hashes in the releases page. 
There are too many files in there. The hashes could be distributed in a different way (if possible).

In the future, we should sign it using a developer account. Today, the user must go to the security page to authorize not only neo-express, but also rocks-db.

File: src/bctklib/persistence/StateServiceStore.cs [hunk 1]
```diff
@@ -268,7 +269,7 @@ static void ValidateProof(UInt256 rootHash, byte[]? proof, (byte[] key, byte[] v
                     Debug.Assert(key.Length == 1);
 
                     var @struct = new VM.Types.Struct() { branchInfo.IndexHash.ToArray(), branchInfo.Index };
-                    return BinarySerializer.Serialize(@struct, 1024 * 1024);
+                    return BinarySerializer.Serialize(@struct, ExecutionEngineLimits.Default with { MaxItemSize = 1024 * 1024 });
```
Comment (lock9) [inline src/bctklib/persistence/StateServiceStore.cs:272]: This is related to Neo 3.6.2, right?

**Review by lock9 — CHANGES_REQUESTED** [review]
It works. 
The releases page became too big.

Do we need to support all of those operating systems? This software will be used by developers, we don't need to make it compatible with servers or mobile devices. 
The hashes are occupying a lot of space, can we distribute it in some other way?

File: src/bctklib/persistence/StateServiceStore.cs (hunk 1 above)
Comment (cschuchardt88) [reply src/bctklib/persistence/StateServiceStore.cs:272]: yes

**Comment by cschuchardt88** [comment]
This is why so many, they for different things, like dockers
- osx-x64 (Minimum OS version is macOS 10.12 Sierra)
- linux-x64 (Most desktop distributions like CentOS, Debian, Fedora, Ubuntu, and derivatives)
- linux-musl-arm64 (Used to build Docker images for 64-bit Arm v8 and minimalistic base images)
- linux-arm64 (Linux distributions running on 64-bit Arm like Ubuntu Server 64-bit on Raspberry Pi Model 3+)

**Comment by Liaojinghui** [comment]
And also, please delete the workflows for other system versions that are not latest. This repo consumes too much github action time, which is limited resource for the whole neo-project.

**Comment by cschuchardt88** [comment]
@Liaojinghui do  you want the `sha256` file removed as well? If we ever do #362 we'll need them

**Comment by Liaojinghui** [comment]
sha256 is small. you can keep the file that you need

**Comment by cschuchardt88** [comment]
whats the artifact feed url?

**Comment by cschuchardt88** [comment]
@Liaojinghui 
For release check out the new output of files please

Also updated the to myget

https://github.com/cschuchardt88/neo-express/releases/tag/3.6.93-preview-g154ca0a8c9

**Comment by lock9** [comment]
Hi @cschuchardt88 , 
Check this release:
https://github.com/neo-project/neo-node/releases/tag/v3.6.2

I think we should do something like it. We don't need to release the Nuget packages here in GitHub if they are available elsewhere.

**Comment by cschuchardt88** [comment]
@lock9 @Liaojinghui
- ill change to remove nuget packages.
- Add portable
- but we are keeping `sha256` for #362 and `win-arm64`?

**Comment by lock9** [comment]
>but we are keeping sha256 for https://github.com/neo-project/neo-express/issues/362 and win-arm64?
They were able to attach the hashes without using a new file. I don't know how they do it but it's possible.

About win-arm64, it seems that this support is [new](https://learn.microsoft.com/en-us/windows/arm/overview). I think we can remove it too.

**Comment by cschuchardt88** [comment]
@lock9 can you add issue about codesigning osx images, 

ill see what @Liaojinghui says on arm64

**Comment by cschuchardt88** [comment]
new folder structure and i updated all links in this pr to this one

https://github.com/cschuchardt88/neo-express/releases/tag/3.6.95-preview-gf56687ff2e

**Comment by cschuchardt88** [comment]
They removed `neo` package `3.6.1` from nuget, we need to update `neo-express`....

```
/home/runner/work/neo-express/neo-express/src/assertions/assertions.csproj : warning NU1603: Neo.Assertions depends on Neo (>= 3.6.1) but Neo 3.6.1 was not found. An approximate best match of Neo 3.6.2 was resolved. 
```

**Comment by Liaojinghui** [comment]
LGTM, merge?

**Comment by cschuchardt88** [comment]
@Liaojinghui yes merge it 👍

**Comment by lock9** [comment]
Hi @Liaojinghui,
I think it shouldn't be merged. There are too many binaries and too many files

**Comment by Liaojinghui** [comment]
please  latest only.

strategy: 
      matrix:
        os: [macos-11, macos-12, macos-13, ubuntu-20.04, ubuntu-22.04, windows-2019, windows-2022]

**Comment by Liaojinghui** [comment]
> Hi @Liaojinghui, I think it shouldn't be merged. There are too many binaries and too many files

You mean on release?

**Comment by lock9** [comment]
If you want to merge it as is, ok, but I think we should try to improve it so it looks more like this in the future https://github.com/neo-project/neo-node/releases/tag/v3.6.2

It deals better with the hashes and platforms.

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -23,33 +23,21 @@ jobs:
   publish:
     needs: [pack, test]
     runs-on: ubuntu-latest
+
     steps:
     - name: Setup .NET
       uses: actions/setup-dotnet@v3
       with:
         dotnet-version: ${{ env.DOTNET_VERSION }}
         source-url: ${{ env.AZURE_ARTIFACTS_FEED_URL }}
```
Comment (Liaojinghui) [inline .github/workflows/push.yml:32]: should this AZURE be updated?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -13,44 +13,53 @@ jobs:
     strategy: 
       matrix:
         os: [macos-11, macos-12, macos-13, ubuntu-20.04, ubuntu-22.04, windows-2019, windows-2022]
```
Comment (Liaojinghui) [inline .github/workflows/test.yml:15]:
```suggestion
        os: [macos-latest, ubuntu-latest, windows-latest]
```

**Comment by Liaojinghui** [comment]
> If you want to merge it as is, ok, but I think we should try to improve it so it looks more like this in the future https://github.com/neo-project/neo-node/releases/tag/v3.6.2
> 
> It deals better with the hashes and platforms.

@lock9 we can discuss about it. but in the meanwhile, please also check the updates in the workflows.

File: .github/workflows/test.yml (hunk 3 above)
Comment (cschuchardt88) [reply .github/workflows/test.yml:15]: check https://github.com/neo-project/neo-express/blob/master/.github/workflows/push.yml do you still want it to publish to github?

**Comment by lock9** [comment]
>@lock9 we can discuss about it. but in the meanwhile, please also check the updates in the workflows.

I don't know why some URLs were changed but I assume that it's because we are going to use MyGet now? I really don't know. Do you know? 😅 

I've tested the osx-arm64 build and it was working fine.  I saw that there are other changes not related to the workflow, but they are related to the Neo 3.6.2 upgrade.

I think that this change is enough for now https://github.com/neo-project/neo-express/pull/358#discussion_r1402828765

**Comment by cschuchardt88** [comment]
@lock9 
These `sha256` files would be used when deploying to `brew` `apt-get` repo or `winget` usage. Where else are we going to storage it for deployment?

**Comment by lock9** [comment]
If it's a requirement, ok. It's better to do separate PRs or at least explain it on the PR.
Note: I don't think we need win-arm64. People do not use it. I think it's fined to keep the hashes if they are used later and must be distributed like this.

File: .github/workflows/push.yml (hunk 2 above)
Comment (cschuchardt88) [reply .github/workflows/push.yml:32]: I can change back, but it was using github repo

Comment (Liaojinghui) [reply .github/workflows/push.yml:32]: nonono, i am just asking, still checking.

**Comment by cschuchardt88** [comment]
Why have `osx-arm64`? My reasoning is that is a standalone app it doesn't require `dotnet` to be installed. Otherwise they can use `portable` if they have `dotnet` installed or the `dotnet tool` command.

File: .github/workflows/push.yml (hunk 2 above)
Comment (cschuchardt88) [reply .github/workflows/push.yml:32]: looking it, doesnt look like we are using the repo at all in this `yml` file

File: .github/workflows/push.yml [hunk 4]
```diff
@@ -23,33 +23,21 @@ jobs:
...
-    - name: Download bctklib
-      uses: actions/download-artifact@v3
-      with:
-        name: bctklib
+        name: packages
         path: ./out
+
     - name: Publish to Azure Artifacts
-      run: dotnet nuget push out/*.nupkg
+      run: dotnet nuget push ./out/*.nupkg --source https://www.myget.org/F/neo/api/v2/package --api-key ${MYGET_TOKEN} --symbol-source https://www.myget.org/F/neo/symbols/api/v2/package --symbol-api-key ${MYGET_TOKEN}
```
Comment (cschuchardt88) [inline .github/workflows/push.yml:43]: Shouldn't the URL be `https://www.myget.org/F/neo/api/v3/index.json` for the source and i get 404 for URL `https://www.myget.org/F/neo/symbols/api/v2/package`

**Comment by lock9** [comment]
> Why have `osx-arm64`? My reasoning is that is a standalone app it doesn't require `dotnet` to be installed. Otherwise they can use `portable` if they have `dotnet` installed or the `dotnet tool` command.

It's because osx-arm64 is used by Apple M1 chips.

**Comment by Liaojinghui** [comment]
ill merge this one, further issues can be fixed in another pr.

**Comment by Liaojinghui** [comment]
@cschuchardt88 publish workflow error.

**Comment by cschuchardt88** [comment]
@Liaojinghui `${MYGET_TOKEN}` is missing from virables

**Comment by cschuchardt88** [comment]
if its added, we may need to do `${{ secrets.MYGET_TOKEN }}`

**Comment by cschuchardt88** [comment]
@Liaojinghui let me know when `${{ secrets.MYGET_TOKEN }}` is added.

**Comment by gsmachado** [comment]
@cschuchardt88 @Liaojinghui guys, where's `linux-musl-arm64`? 😢 

For example, `alpine` uses musl as its default C library. I think it's super useful besides the usual `linux-arm64`.

**Comment by cschuchardt88** [comment]
@gsmachado ya i think so too. but have people that think overwise. I'll add it back, because no cares this about this repo.

**Comment by gsmachado** [comment]
> @gsmachado ya i think so too. but have people that think overwise. I'll add it back, because no cares this about this repo.

I agree 100% with your comment here: https://github.com/neo-project/neo-express/pull/358#issuecomment-1820118613

I suggest sticking to your original plan. Developers will know what to choose from (answering the argument that there are "too many files"). The worst thing is to exclude important developer software/hardware, e.g., `alpine` docker containers, or raspberry pi (which I personally use to develop and run docker hosts).

Add it back! 😄

**Pull request merged**